045-kubernetes-operator/proposal.md

```diff
@@ -0,0 +1,315 @@
+
+    PostgreSQL Configuration:
+          --postgres-host=                                           
+          --postgres-port=                                           
+          --postgres-socket=                                         
+          --postgres-user=                                           
+          --postgres-password=                                       
+          --postgres-sslmode=[disable|require|verify-ca|verify-full] 
+          --postgres-ca-cert=                                        
+          --postgres-client-cert=                                    
+          --postgres-client-key=                                     
+          --postgres-connect-timeout=                                
+          --postgres-database=                                       
+```
+
+While we try our best to never need to perform a rollback, that can eventually
+be necessary. This is especially true for the Concourse team itself, where we
+want to have Concourse being continuously deployed, while making bad deployments
+"no-factor" situations.
+
+
+### other operational aspects that could be automated
+
+Aside from those two cases, there are other (probably less interesting) ones
+that could also be made better.
+
+
+#### db encryption key rotation
+
+To rotate the database encryption key, one needs specify the right set of flags
+to `concourse web` so that it can go from the old encryption to the new one.
+
+With a schedule to rotate those keys, we could automate such process and provide
+the right flags.
+
+See [rotating the encryption key].
+
+
+#### maintenance mode
+
+With the introduction of component tracking (see [#4534]), it's now possible to
+tweak the intervals at which certain components of Concourse run. 
+
+This means that in the case of *fire* (or some form of maintenance in the
+infrastructure), one can, for instance, disable a given set of components (say,
+scheduling of builds and checking or resources) altogether, rendering the
+installation able to accept inputs from the users, while still sending no load
+to the workers (which could even not exist at the moment).
+
+Tying that to the recently added ability of setting "messages for the users"
+([#5019]), one could transition a given installation to "maintenance mode" by
+switching a flag on the Kubernetes object that represents that installation
+(which could be done in a "gitops fashion").
+
+
+[#4534]: https://github.com/concourse/concourse/issues/4534
+[#5019]: https://github.com/concourse/concourse/pull/5019
+
+
+# Proposal
+
+Have a Concourse [Kubernetes Operator] to deal with the whole deployment of
+Concourse on Kubernetes installations.
+
+### creation
+
+At a high-level, one should be able to submit to a Kubernetes cluster an
+object that represents a Concourse installation:
+
+
+```yaml
+apiVersion: concourse.concourse-ci.org/v1
+kind: Concourse
+metadata:
+  name: hush-house
+spec:
+  replicas: 2
```

Comment:
Am I correct to infer that this CRD only represents ATC's, and isn't responsible for creating workers? Or is this looking to a point beyond having worker pods, where the cluster itself is a "worker"?

I'm not too familiar with how the [k8s worker experiment](https://github.com/concourse/concourse/pull/5223) you created works, but it may be the case that with "k8s workers", the ATC creates pods for steps/resource checks (and there's no dedicated "worker" pods)?